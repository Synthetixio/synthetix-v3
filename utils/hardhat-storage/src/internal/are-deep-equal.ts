import { deepEqual } from 'node:assert';

export function areDeepEqual(a: unknown, b: unknown) {
  try {
    deepEqual(a, b);
    return true;
  } catch {
    return false;
  }
}
