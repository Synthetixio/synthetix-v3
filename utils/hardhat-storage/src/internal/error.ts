import { render } from './render';

import type { ASTNode, AssemblyItem, BaseASTNode } from '@solidity-parser/parser/src/ast-types';

export class ValidationError extends Error {}

/**
 * Create a ValidationError with a custom stack trace generated based on the
 * given AST nodes, and print the solidity path stack trace.
 */
export function createError({
  message,
  sourceName,
  nodes = [],
}: {
  message: string;
  sourceName: string;
  nodes: (ASTNode | AssemblyItem)[];
}) {
  const err = new ValidationError(message);
  const [title] = err.stack!.split('\n');
  const stack: string[] = [];

  try {
    for (const node of nodes) {
      const title = _renderTitle(node);
      if (!title) continue;
      stack.unshift(`    at ${title} (${sourceName}${_renderLoc(node)})`);
    }
  } catch (err) {
    console.warn('Could not render stack trace.', err);
  }

  stack.unshift(title);
  err.stack = stack.join('\n');
  return err;
}

function _renderTitle(node: ASTNode | AssemblyItem) {
  switch (node.type) {
    case 'ContractDefinition':
      return `${node.kind} ${node.name}`;
    case 'VariableDeclaration':
      return render(node);
    default:
      return typeof (node as { name: string }).name === 'string'
        ? (node as { name: string }).name
        : node.type;
  }
}

function _renderLoc(node: BaseASTNode) {
  if (!node.loc) return '';
  const loc = [`:${node.loc.start.line}`];
  if (node.loc.start.column) loc.push(`:${node.loc.start.column}`);
  return loc.join('');
}
