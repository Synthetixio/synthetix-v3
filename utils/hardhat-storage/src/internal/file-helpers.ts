import fs from 'node:fs/promises';
import path from 'node:path';

export async function readFile(filepath: string) {
  return fs.readFile(filepath, { encoding: 'utf8' }).then((res) => res.toString());
}

export async function readFileSafe(filepath: string) {
  try {
    const result = await readFile(filepath);
    return result;
  } catch (err: unknown) {
    if (_isNodeError(err) && err.code === 'ENOENT') return;
    throw err;
  }
}

function _isNodeError(err: unknown): err is Error & { code?: string } {
  return err instanceof Error && typeof (err as Error & { code?: string }).code === 'string';
}

export async function writeFile(target: string, content: string) {
  await fs.mkdir(path.dirname(target), { recursive: true });
  return fs.writeFile(target, content);
}

export async function readJsonFile<T = unknown>(filepath: string) {
  const content = await readFile(filepath);
  return JSON.parse(content) as T;
}

export async function readJsonFileSafe<T = unknown>(filepath: string) {
  const content = await readFileSafe(filepath);
  return content ? (JSON.parse(content) as T) : undefined;
}

export async function writeJsonFile(target: string, content: unknown) {
  const data = typeof content === 'string' ? content : JSON.stringify(content, null, 2);
  return writeFile(target, data + '\n');
}
