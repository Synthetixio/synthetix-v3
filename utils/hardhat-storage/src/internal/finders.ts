import * as parser from '@solidity-parser/parser';

import type {
  ASTNode,
  ASTNodeTypeString,
  ContractDefinition,
  SourceUnit,
} from '@solidity-parser/parser/src/ast-types';

export type ASTMap<U> = { [K in ASTNodeTypeString]: U extends { type: K } ? U : never };
export type ASTTypeMap = ASTMap<ASTNode>;

// parse.visit() function but compatible with multiple astNodes and better types interface
function _visit<T extends ASTNodeTypeString>(
  astNodes: ASTNode | ASTNode[],
  nodeTypes: T | T[],
  visitorFn: (node: ASTTypeMap[T]) => false | void
) {
  const visitor: { [nodeType: string]: (node: ASTTypeMap[T]) => false | void } = {};
  let cancelled = false;

  const astNodesArr = Array.isArray(astNodes) ? astNodes : [astNodes];
  const nodeTypesArr = Array.isArray(nodeTypes) ? nodeTypes : [nodeTypes];

  const _visitorFn = (node: ASTTypeMap[T]) => {
    const res = visitorFn(node);
    if (res === false) cancelled = true;
  };

  for (const nodeType of nodeTypesArr) {
    visitor[nodeType] = _visitorFn;
  }

  for (const astNode of astNodesArr) {
    if (cancelled) break;
    parser.visit(astNode, visitor);
  }
}

export function findAll<T extends ASTNodeTypeString>(
  astNodes: ASTNode | ASTNode[],
  nodeTypes: T | T[],
  filterFn: (node: ASTTypeMap[T]) => boolean = () => true
) {
  const results: ASTTypeMap[T][] = [];

  _visit(astNodes, nodeTypes, (node) => {
    if (filterFn(node)) results.push(node);
  });

  return results;
}

export function findOne<T extends ASTNodeTypeString>(
  astNodes: ASTNode | ASTNode[],
  nodeTypes: T | T[],
  filterFn: (node: ASTTypeMap[T]) => boolean = () => true
): ASTTypeMap[T] | undefined {
  let result: ASTTypeMap[T] | undefined = undefined;

  _visit(astNodes, nodeTypes, (node) => {
    if (filterFn(node)) {
      result = node;
      return false;
    }
  });

  return result;
}

export function findChildren<T extends ASTNodeTypeString>(
  astNode: SourceUnit,
  nodeTypes: T | T[],
  filterFn: (node: ASTTypeMap[T]) => boolean = () => true
) {
  const types: ASTNodeTypeString[] = Array.isArray(nodeTypes) ? nodeTypes : [nodeTypes];
  return astNode.children.filter(
    (node: ASTNode) => types.includes(node.type) && filterFn(node as ASTTypeMap[T])
  );
}

export function findContract(
  astNode: SourceUnit,
  contractName: string,
  filterFn: (node: ContractDefinition) => boolean = () => true
) {
  return astNode.children.find(
    (node) => node.type === 'ContractDefinition' && node.name === contractName && filterFn(node)
  ) as ContractDefinition | undefined;
}

export function findContractStrict(
  astNode: SourceUnit,
  contractName: string,
  filterFn: (node: ContractDefinition) => boolean = () => true
) {
  const contractNode = findContract(astNode, contractName, filterFn);

  if (!contractNode) {
    throw new Error(`Contract with name "${contractName}" not found`);
  }

  return contractNode;
}

export function getCanonicalImportedSymbolName(
  astNode: ASTNode,
  symbolName: string
): [string, string] | undefined {
  for (const imp of findAll(astNode, 'ImportDirective')) {
    if (imp.symbolAliases) {
      for (const [canonicalName, alias] of imp.symbolAliases) {
        if (canonicalName === symbolName || alias === symbolName) {
          return [imp.path, canonicalName];
        }
      }
    }
  }
}
