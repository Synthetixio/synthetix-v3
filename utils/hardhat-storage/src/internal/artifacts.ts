import path from 'node:path';
import * as parser from '@solidity-parser/parser';
import { ASTNodeTypeString, ContractDefinition } from '@solidity-parser/parser/src/ast-types';
import { GetArtifactFunction, StorageArtifact } from '../types';
import { readFileSafe } from './file-helpers';
import {
  ASTTypeMap,
  findAll,
  findContractStrict,
  findOne,
  getCanonicalImportedSymbolName,
} from './finders';
import { ensureTrailingSlash, isExplicitRelativePath, removeBasePath } from './path-helpers';

export async function readArtifact(projectRoot: string, sourceName: string, sourcePrefix = '') {
  if (sourcePrefix && !sourceName.startsWith(ensureTrailingSlash(sourcePrefix))) {
    throw new Error(`Invalid sourcePrefix "${sourcePrefix}" for "${sourceName}"`);
  }

  const sourcePath = sourcePrefix
    ? sourceName.slice(ensureTrailingSlash(sourcePrefix).length)
    : sourceName;

  const sourceFullPath = path.resolve(projectRoot, sourcePath);
  const sourceCode = await readFileSafe(sourceFullPath);

  if (sourceCode === undefined) {
    throw new Error(`Could not find "${sourceName}" at "${sourceFullPath}"`);
  }

  const sourceUnit = parser.parse(sourceCode, { loc: true });

  // Modify all import directives to start from `projectRoot`
  _normalizeImportDirectives(projectRoot, sourceFullPath, sourceUnit, sourcePrefix);

  return {
    sourceName,
    ast: sourceUnit,
  } satisfies StorageArtifact;
}

function _normalizeImportDirectives(
  projectRoot: string,
  sourceFullPath: string,
  sourceUnit: StorageArtifact['ast'],
  sourcePrefix: string
) {
  for (const node of findAll(sourceUnit, 'ImportDirective')) {
    if (!isExplicitRelativePath(node.path)) continue;
    const sourceName = removeBasePath(
      projectRoot,
      path.resolve(path.dirname(sourceFullPath), node.path)
    );
    const target = sourcePrefix ? path.join(sourcePrefix, sourceName) : sourceName;
    node.path = target;
    node.pathLiteral.value = target;
    node.pathLiteral.parts = [target];
  }
}

export async function findNodeReferenceWithArtifact<T extends ASTNodeTypeString>(
  getArtifact: GetArtifactFunction,
  nodeTypes: T | T[],
  artifact: StorageArtifact,
  nodePath: string
): Promise<[StorageArtifact, ASTTypeMap[T]]> {
  // Handle children nodes, e.g.: ContractName.StrcutName
  if (nodePath.includes('.')) {
    const names = nodePath.split('.');

    if (names.length > 2) {
      throw new Error(
        `Cannot handle node references with more than 1 parent in "${nodePath}" from "${artifact.sourceName}"`
      );
    }

    const [parentNodeName, childNodeName] = names;

    const [parentArtifact, parentNode] = await findNodeReferenceWithArtifact(
      getArtifact,
      nodeTypes,
      artifact,
      parentNodeName
    );

    const childNode = findOne(
      parentNode,
      nodeTypes,
      (node) => (node as { name: string }).name === childNodeName
    );

    if (!childNode) {
      throw new Error(
        `Could not find node with name "${nodePath}" from "${parentArtifact.sourceName}"`
      );
    }

    return [parentArtifact, childNode];
  }

  // Check if it's defined on the same file
  const localNode = findOne(
    artifact.ast,
    nodeTypes,
    (node) => (node as { name: string }).name === nodePath
  );

  if (localNode) return [artifact, localNode];

  const importDirective = getCanonicalImportedSymbolName(artifact.ast, nodePath);

  // If we have the real name and where it was imported from, look for it there
  if (importDirective) {
    const [importSourceName, canonicalNodeName] = importDirective;
    const importedArtifact = await getArtifact(importSourceName);
    const foundNode = findOne(
      importedArtifact.ast,
      nodeTypes,
      (node) => (node as { name: string }).name === canonicalNodeName
    );
    if (foundNode) return [importedArtifact, foundNode];
  } else {
    // if not, recursively on all the imported files
    const importedSourceNames = findAll(artifact.ast, 'ImportDirective').map((node) => node.path);

    // loop protection: prevents infinite loop in case recursion of file imports causes that
    const doneSourceNames = new Set(importedSourceNames);

    let importSourceName;
    while ((importSourceName = importedSourceNames.pop())) {
      const importedArtifact = await getArtifact(importSourceName);
      const foundNode = findOne(
        importedArtifact.ast,
        nodeTypes,
        (node) => (node as { name: string }).name === nodePath
      );
      if (foundNode) return [importedArtifact, foundNode];

      // recursively add all import directives from this file
      for (const d of findAll(importedArtifact.ast, 'ImportDirective').map((node) => node.path)) {
        if (!doneSourceNames.has(d)) {
          importedSourceNames.push(d);
          doneSourceNames.add(d);
        }
      }
    }
  }

  throw new Error(`Could not find node with name "${nodePath}" from "${artifact.sourceName}"`);
}

export async function findContractTree(
  getArtifact: GetArtifactFunction,
  artifact: StorageArtifact,
  contractName: string
) {
  const contractNode = findContractStrict(artifact.ast, contractName);

  const contractNodes: ContractDefinition[] = [contractNode];

  for (const baseContract of contractNode.baseContracts) {
    const [inheritedArtifact, inheritedContract] = await findNodeReferenceWithArtifact(
      getArtifact,
      'ContractDefinition',
      artifact,
      baseContract.baseName.namePath
    );

    const importedContractTree = await findContractTree(
      getArtifact,
      inheritedArtifact,
      inheritedContract.name
    );

    contractNodes.push(...importedContractTree);
  }

  return contractNodes;
}
