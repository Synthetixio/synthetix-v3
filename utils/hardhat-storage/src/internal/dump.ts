import { findAll } from '@synthetixio/core-utils/utils/ast/finders';
import { clone } from '@synthetixio/core-utils/utils/misc/clone';
import { SourceUnit } from 'solidity-ast/types';
import { iterateContracts, iterateSlotAssignments } from './iterators';
import { render } from './render';

/**
 * Generate a single solidity file including all the given contracts but only
 * rendering its storage defintions.
 */
export async function dumpStorage(
  astNodes: SourceUnit[],
  version?: string,
  license = 'UNLICENSED'
) {
  if (!Array.isArray(astNodes) || !astNodes.length) {
    throw new Error('No solidity files found');
  }

  const result = [
    `// SPDX-License-Identifier: ${license}`,
    version ? `pragma solidity ${version};` : _renderPragmaDirective(astNodes),
    '',
  ];

  for (const [sourceUnit, contractNode] of iterateContracts(astNodes)) {
    const sourceName = sourceUnit.absolutePath;
    const contractName = contractNode.name;
    const fqName = `${sourceName}:${contractName}`;

    const resultNode = clone(contractNode);

    const constDeclarations = findAll(
      contractNode,
      'VariableDeclaration',
      (node) => node.mutability === 'constant'
    );
    const enumDefinitions = findAll(contractNode, 'EnumDefinition');
    const structDefinitions = findAll(contractNode, 'StructDefinition');
    const slotAssignments = [...iterateSlotAssignments([sourceUnit])].map(
      ([, , functionNode]) => functionNode
    );

    // Filter all the contract nodes to only include Structs and Storage Slot Definitions
    resultNode.nodes = [
      ...constDeclarations,
      ...enumDefinitions,
      ...structDefinitions,
      ...slotAssignments,
    ];

    if (!resultNode.nodes.length) continue;

    // Render the contract only including storage definitions
    const contract = render(resultNode);

    result.push(`// @custom:artifact ${fqName}`, contract, '');
  }

  return result.join('\n');
}

function _renderPragmaDirective(sourceUnits: SourceUnit[]) {
  const sourceUnit = sourceUnits[sourceUnits.length - 1];
  // TODO: calculate the best solc version based on all the files, instead of using
  // the one from the last file
  for (const node of findAll(sourceUnit, 'PragmaDirective')) return render(node);
}
