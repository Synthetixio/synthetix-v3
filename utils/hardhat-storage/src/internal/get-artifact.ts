import fs from 'node:fs/promises';
import path from 'node:path';
import * as parser from '@solidity-parser/parser';
import { StorageArtifact } from '../types';
import { findAll } from './finders';

export async function getArtifact(projectRoot: string, sourceName: string, contractName: string) {
  const sourceFullPath = path.resolve(projectRoot, sourceName);
  const source = await fs.readFile(sourceFullPath, { encoding: 'utf8' });
  const sourceUnit = parser.parse(source, { loc: true });

  // Modify all import directives to start from `projectRoot`
  _normalizeImportDirectives(projectRoot, sourceFullPath, sourceUnit);

  return {
    sourceName,
    contractName,
    ast: sourceUnit,
  } satisfies StorageArtifact;
}

function _isExplicitRelativePath(sourceName: string): boolean {
  return sourceName.startsWith('./') || sourceName.startsWith('../');
}

function _normalizeImportDirectives(
  projectRoot: string,
  sourceFullPath: string,
  sourceUnit: StorageArtifact['ast']
) {
  for (const node of findAll(sourceUnit, 'ImportDirective')) {
    if (!_isExplicitRelativePath(node.path)) continue;
    const target = _removeBasePath(projectRoot, sourceFullPath);
    node.path = target;
    node.pathLiteral.value = target;
    node.pathLiteral.parts = [target];
  }
}

function _removeBasePath(basePath: string, fullPath: string) {
  basePath = path.normalize(basePath);
  fullPath = path.normalize(fullPath);

  if (!basePath.endsWith(path.sep)) basePath += path.sep;

  if (!fullPath.startsWith(basePath)) {
    throw new Error(`The path "${fullPath}" is not inside "${fullPath}"`);
  }

  return fullPath.substring(basePath.length);
}
