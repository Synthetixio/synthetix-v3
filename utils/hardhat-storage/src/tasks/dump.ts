import path from 'node:path';
import logger from '@synthetixio/core-utils/utils/io/logger';
import { task } from 'hardhat/config';
import { dumpStorage } from '../internal/dump';
import { readJsonFileSafe, writeJsonFile } from '../internal/file-helpers';
import { logInChunks } from '../internal/log-in-chunks';
import {
  TASK_STORAGE_DUMP,
  SUBTASK_GET_ARTIFACTS,
  SUBTASK_VALIDATE_CONTRACTS,
  SUBTASK_VERIFY_CONTRACTS,
} from '../task-names';
import { StorageDump } from '../types';

interface Params {
  output: string;
  noValidate: boolean;
  noVerify: boolean;
  quiet: boolean;
  log: boolean;
}

task(TASK_STORAGE_DUMP, 'Dump storage slots to a file')
  .addOptionalParam(
    'output',
    'Storage dump output file relative to the root of the project',
    'storage.dump.json'
  )
  .addFlag('noValidate', 'Do not perform static validations on contracts before generating')
  .addFlag('noVerify', 'Do not verify storage mutations before replacing dump')
  .addFlag('quiet', 'only emit errors to the console')
  .addFlag('log', 'log json result to the console')
  .setAction(async (params: Params, hre) => {
    const { output, noValidate, noVerify, log, quiet } = params;

    const now = Date.now();
    const target = path.resolve(hre.config.paths.root, output);

    await hre.run('compile');

    const { contracts, getArtifact } = await hre.run(SUBTASK_GET_ARTIFACTS);

    if (!noValidate) {
      await hre.run(SUBTASK_VALIDATE_CONTRACTS, { contracts, getArtifact });
    }

    const dump = await dumpStorage({ contracts, getArtifact });

    if (!noVerify) {
      const prev = await readJsonFileSafe<StorageDump>(target);
      await hre.run(SUBTASK_VERIFY_CONTRACTS, {
        curr: dump,
        prev,
        quiet: log || quiet,
      });
    }

    if (output && dump) {
      await writeJsonFile(target, dump);
    }

    if (log) {
      if (dump) logInChunks(dump);
    } else if (!quiet) {
      logger.success(`Storage dump finished in ${Date.now() - now}ms`);
    }

    return dump;
  });
