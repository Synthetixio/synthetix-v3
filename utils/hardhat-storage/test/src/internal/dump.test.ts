import path from 'node:path';
import { compileSolidityFolder } from '@synthetixio/core-utils/utils/solidity/compiler';
import { dumpStorage } from '../../../src/internal/dump';

const version = '0.8.17';

describe('internal/dump.ts', function () {
  jest.setTimeout(120000);

  it('single contract with state variables, contracts and structs', async function () {
    const astNodes = await compileSolidityFolder({
      version,
      rootDir: path.resolve(__dirname, '..', '..', 'fixtures'),
      sources: 'ExampleContract.sol',
    });

    const result = await dumpStorage(astNodes, version);
    expect(result).toMatchSnapshot();
  });

  it('sample-project contract with storage and interface', async function () {
    const astNodes = await compileSolidityFolder({
      version,
      rootDir: path.resolve(__dirname, '..', '..', '..', '..', 'sample-project'),
      sources: [
        'contracts/storage/GlobalStorage.sol',
        'contracts/interfaces/ISomeModule.sol',
        'contracts/modules/SomeModule.sol',
      ],
    });

    const result = await dumpStorage(astNodes, version);
    expect(result).toMatchSnapshot();
  });
});
