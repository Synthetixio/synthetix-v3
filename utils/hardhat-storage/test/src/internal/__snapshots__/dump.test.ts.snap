// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`internal/dump.ts sample-project contract with storage and interface 1`] = `
"// SPDX-License-Identifier: UNLICENSED
pragma solidity 0.8.17;

// @custom:artifact contracts/storage/GlobalStorage.sol:GlobalStorage
contract GlobalStorage {
    bytes32 private constant _SLOT_GLOBAL_STORAGE = keccak256(abi.encode("io.synthetix.sample-project.Global"));
    struct GlobalStore {
        uint value;
        uint someValue;
    }
    function _globalStore() internal pure returns (GlobalStore storage store) {
        bytes32 s = _SLOT_GLOBAL_STORAGE;
        assembly {
            store.slot := s
        }
    }
}
"
`;

exports[`internal/dump.ts single contract with state variables, contracts and structs 1`] = `
"// SPDX-License-Identifier: UNLICENSED
pragma solidity 0.8.17;

// @custom:artifact ExampleContract.sol:ExampleContract
contract ExampleContract {
    bytes32 private constant _SLOT_EXAMPLE_CONTRACT = keccak256(abi.encode("io.synthetix.hardhat-storage.Example"));
    uint128 public constant SOME_CONSTANT = 445;
    uint128 public constant SOME_CASTED_CONSTANT = uint128(SOME_CONSTANT);
    address private constant SOME_ADDRESS = 0xe27454c382e79a1876096B691ef4b52747B7097D;
    enum NodeType {
        NONE,
        REDUCER,
        EXTERNAL,
        CHAINLINK,
        PYTH
    }
    struct Data {
        address owner;
        mapping(address => uint128) permissions;
        NodeType nodeType;
    }
    function load() internal pure returns (Data storage store) {
        bytes32 s = _SLOT_EXAMPLE_CONTRACT;
        assembly {
            store.slot := s
        }
    }
}
"
`;
